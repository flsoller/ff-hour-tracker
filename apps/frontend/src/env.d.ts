/// <reference types="vite/client" />

declare module "*.vue" {
  import { DefineComponent } from "vue";
  const component: DefineComponent<object, object, unknown>;
  export default component;
}

interface ImportMetaEnv {
  readonly VITE_API_URL: string;
}

interface ImportMeta {
  readonly env: ImportMetaEnv;
}
