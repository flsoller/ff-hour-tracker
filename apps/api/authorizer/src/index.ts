import {
  Context,
  APIGatewayProxyEventV2,
  APIGatewaySimpleAuthorizerWithContextResult,
} from "aws-lambda";
import { verifyAccessToken } from "./services/jwt";
import { getActiveUser } from "./services/user";
import { UserContext } from "./types/context.type";
import { logger } from "@hour-tracker/logger";

export const handler = async (
  event: APIGatewayProxyEventV2
): Promise<APIGatewaySimpleAuthorizerWithContextResult<UserContext | {}>> => {
  const token = event.headers.authorization?.split("Bearer ")[1];

  // Deny access when no token
  if (!token) {
    logger.warn("No token provided");
    return {
      isAuthorized: false,
      context: {},
    };
  }

  const { userId, organizationId } = verifyAccessToken(token);

  // Deny access when no userId, organizationId can be obtained from token
  if (!userId || !organizationId) {
    logger.warn("Invalid token provided");
    return {
      isAuthorized: false,
      context: {},
    };
  }

  const { isActive, user } = await getActiveUser(userId, organizationId);
  const isAuthorized = isActive && !!user;

  logger.info({ isActive, user }, "User request evaluated");

  return {
    isAuthorized,
    context: isAuthorized
      ? {
          userId,
          organizationId,
          role: user?.role ?? "",
        }
      : {},
  };
};
