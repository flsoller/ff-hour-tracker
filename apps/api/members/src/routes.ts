import { getMembersPaginated } from "./v1/paginated/get-members.route";
import { V1_GET_PAGINATED_ROUTE_KEY } from "./v1/paginated/get-members.type";

const ROUTE_KEYS = {
  V1_GET_PAGINATED_ROUTE_KEY,
};

export const ROUTE_HANDLING_MAP = {
  [ROUTE_KEYS.V1_GET_PAGINATED_ROUTE_KEY]: getMembersPaginated,
};

export function getRouteHandler(routeKey: string) {
  return ROUTE_HANDLING_MAP[routeKey] ?? null;
}
