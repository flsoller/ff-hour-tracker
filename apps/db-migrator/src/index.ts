import { logger } from "@hour-tracker/logger";
import { drizzle } from "drizzle-orm/postgres-js";
import { migrate } from "drizzle-orm/postgres-js/migrator";
import postgres from "postgres";
import { CONNECTION_STRING } from "./config";

const migrationClient = postgres(CONNECTION_STRING, { max: 1 });

async function migrateDb() {
  logger.info("Initiating Migration");
  await migrate(drizzle(migrationClient), {
    migrationsFolder: __dirname + "/migrations",
    migrationsTable: "migrations",
    migrationsSchema: "hour_tracker_migrations",
  });
}

migrateDb()
  .then(() => {
    logger.info("Database Migration Completed");
    process.exit();
  })
  .catch((e) => {
    logger.info("Error during migration", e);
    process.exit(1);
  });
