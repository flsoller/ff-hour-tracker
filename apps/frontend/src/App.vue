<template>
  <div class="app-container">
    <Toast />
    <DynamicDialog />
    <Sidebar v-if="!$route.meta.hideSidebar" />
    <div class="app-container__content">
      <router-view />
    </div>
  </div>
</template>

<script setup lang="ts">
import Sidebar from "@/components/AppSidebar.vue";
import DynamicDialog from "primevue/dynamicdialog";
import Toast from "primevue/toast";
</script>

<style lang="scss">
@import "@/styles/base/normalize.scss";
@import "@/styles/theme.scss";

.app-container {
  display: flex;
  height: 100vh;
  width: 100vw;

  &__content {
    width: 100%;
  }
}
</style>
