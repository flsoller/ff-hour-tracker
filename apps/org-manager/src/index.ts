import { Context } from "aws-lambda";

export const handler = async (
  event: unknown,
  context: Context
): Promise<void> => {
  try {
    console.log("Lambda invoked with:", event);
  } catch (error) {
    console.error("Error during");
  }
};
