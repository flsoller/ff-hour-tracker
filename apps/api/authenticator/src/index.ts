import { Context, APIGatewayProxyEventV2 } from "aws-lambda";
import { handleError } from "@hour-tracker/lambda-api/errorHandler";
import { NotFoundError } from "@hour-tracker/lambda-api/errors";
import { getRouteHandler, ROUTE_HANDLING_MAP } from "./routes";

export const handler = async (
  event: APIGatewayProxyEventV2,
  context: Context
): Promise<unknown> => {
  try {
    const routeHandler = getRouteHandler(event.requestContext.routeKey);
    if (!routeHandler) {
      throw new NotFoundError();
    }
    return await routeHandler(event);
  } catch (error) {
    return handleError(error);
  }
};
