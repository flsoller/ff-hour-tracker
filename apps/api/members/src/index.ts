import { handleError } from "@hour-tracker/lambda-api/errorHandler";
import { NotFoundError } from "@hour-tracker/lambda-api/errors";
import { getRouteHandler } from "./routes";
import { APIRequest } from "./common/types/request.type";

export const handler = async (event: APIRequest): Promise<unknown> => {
  console.log(event);

  try {
    const routeHandler = getRouteHandler(event.routeKey);
    if (!routeHandler) {
      throw new NotFoundError();
    }
    return await routeHandler(event);
  } catch (error) {
    return handleError(error);
  }
};
