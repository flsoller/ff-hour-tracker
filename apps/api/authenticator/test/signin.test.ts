import { Context } from "aws-lambda";
import {
  SignInEventBody,
  SignInResponse,
} from "../src/common/types/signin.types";
import { handler } from "../src";
import {
  createAuthenticatorEvent,
  createOrganizations,
  createUserForOrganization,
} from "./helpers/test-data";
import { logger } from "@hour-tracker/logger";
import { db } from "@hour-tracker/db";

describe("SignIn", () => {
  let orgId: string;

  beforeEach(async () => {
    const [organization] = await createOrganizations();
    orgId = organization!.id;
  });

  describe("handleSignIn", () => {
    it("should return access token on successful login requests", async () => {
      const [user] = await createUserForOrganization(orgId);
      const event = createAuthenticatorEvent<SignInEventBody>(
        {
          emailAddress: user!.emailAddress,
          password: "12345",
        },
        "/auth/signin",
        "POST"
      );
      const res = (await handler(event, {} as Context)) as SignInResponse;
      expect(res.accessToken).toEqual(expect.any(String));
    });

    it("should return a 400 error when user does not exist", async () => {
      const event = createAuthenticatorEvent<SignInEventBody>(
        {
          emailAddress: "not-here@ghost.com",
          password: "abc",
        },
        "/auth/signin",
        "POST"
      );
      expect(await handler(event, {} as Context)).toMatchSnapshot();
    });

    it("should return a 400 error when password is incorrect", async () => {
      const [user] = await createUserForOrganization(orgId);
      const event = createAuthenticatorEvent<SignInEventBody>(
        {
          emailAddress: user!.emailAddress,
          password: "abc",
        },
        "/auth/signin",
        "POST"
      );
      expect(await handler(event, {} as Context)).toMatchSnapshot();
    });

    it("should return a 500 error when there is an internal server error", async () => {
      const loggerSpy = jest.spyOn(logger, "error");
      jest.spyOn(db, "select").mockImplementationOnce(() => {
        throw new Error("Database connection failed");
      });
      const event = createAuthenticatorEvent<SignInEventBody>(
        {
          emailAddress: "not-here@ghost.com",
          password: "abc",
        },
        "/auth/signin",
        "POST"
      );
      const response = await handler(event, {} as Context);
      expect(response).toMatchObject({
        statusCode: 500,
        body: JSON.stringify({
          message: "API error while processing request",
          details: [],
        }),
      });
      expect(loggerSpy).toHaveBeenCalledWith(
        expect.objectContaining({
          error: expect.any(Error),
        }),
        "InternalServerError"
      );
    });
  });
});
