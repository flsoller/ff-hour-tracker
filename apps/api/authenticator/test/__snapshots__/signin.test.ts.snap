// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn handleSignIn should return a 400 error when password is incorrect 1`] = `
{
  "body": "{"message":"InvalidInformation","details":[]}",
  "statusCode": 400,
}
`;

exports[`SignIn handleSignIn should return a 400 error when user does not exist 1`] = `
{
  "body": "{"message":"InvalidInformation","details":[]}",
  "statusCode": 400,
}
`;
